src/modules/sync/lib/convert-tree-to-sync-options.ts

```diff
@@ -9,78 +9,130 @@ type PushOptionsWithSelections = {
 		plugins?: string[];
 		themes?: string[];
 		uploads?: string[];
+		contents?: string[];
 	};
 };
 
-const isSyncOption = ( value: string ): value is SyncOption => {
-	return Object.keys( SYNC_OPTIONS ).includes( value );
+const STANDARD_CATEGORIES = [
+	{ prefix: 'plugins/', key: 'plugins' as const, option: SYNC_OPTIONS.plugins },
+	{ prefix: 'themes/', key: 'themes' as const, option: SYNC_OPTIONS.themes },
+	{ prefix: 'uploads/', key: 'uploads' as const, option: SYNC_OPTIONS.uploads },
+] as const;
+
+const CONTENTS_ROOTS = [ 'mu-plugins', 'fonts', 'languages' ] as const;
+
+const normalizePath = ( path: string ): string =>
+	path.replace( /^\/?wp-content\//, '' ).replace( /^\/+|\/+$/g, '' );
+
+const traverseSelected = (
+	nodes: TreeNode[] | undefined,
+	visit: ( node: TreeNode ) => void
+): void => {
+	if ( ! nodes?.length ) return;
+	for ( const node of nodes ) {
+		if ( node.checked ) {
+			visit( node );
+		} else if ( node.indeterminate && node.children?.length ) {
+			traverseSelected( node.children, visit );
+		}
+	}
+};
+
+const collectPathIds = ( nodes: TreeNode[] | undefined ): string[] => {
+	const out: string[] = [];
+	traverseSelected( nodes, ( node ) => {
+		if ( node.pathId ) out.push( node.pathId );
+	} );
+	return out;
+};
+
+type Categories = {
+	plugins: Set< string >;
+	themes: Set< string >;
+	uploads: Set< string >;
+	contents: Set< string >;
 };
 
-const collectPathIds = ( nodes: TreeNode[], pathIds: string[] = [] ): string[] => {
-	nodes.forEach( ( node ) => {
-		if ( node.checked && node.pathId ) {
-			pathIds.push( node.pathId );
-		} else if ( node.indeterminate && node.children ) {
-			collectPathIds( node.children, pathIds );
+const categorizeSelectedPathsInPlace = (
+	nodes: TreeNode[] | undefined,
+	categories: Categories
+): void => {
```

Comment:
```suggestion
const categorizeSelectedPathsInPlace = ( nodes: TreeNode[] | undefined ) => {
	const categories: Categories = {
		plugins: new Set< string >(),
		themes: new Set< string >(),
		uploads: new Set< string >(),
		contents: new Set< string >(),
	};
```
Let's construct the `categories` object inside this function instead. There's no reason to accept it as an argument.

Reply:
https://github.com/Automattic/studio/pull/1940/commits/76fc204d9de4852bbf5c0511a16b572367b48c7b